### ISSUE TITLE
I started to write an OpenGL backend

### ISSUE THREAD

**Issue by DavidGOrtega** [tensorflow/tfjs#811]
Hi guys,

yesterday I started to mock up a new backend based on openGL to access any GPU on Mac and Windows. It sounds quite logical to me since tfjs has backend WebGL, generating bindings for openGL would be a better option to address any possible non CUDA GPU in desktop specially the cheapo ones in laptops apart of Intel and AMD.

You must have thought about it for sure, and that makes me doubt if this is possible. Any reason the project would not be doable?

Please any feedback before I sacrifice some time on this would make me very happy.

**Comment by nkreeger** [comment]
Stay tuned - we're looking at providing support for this. This isn't as easy as "run on the system OpenGL". WebGL doesn't fully reflect OpenGL ES 2.0 - it is closer to ES 3.0 but still has a variety of issues.

**Comment by DavidGOrtega** [comment]
@nkreeger awesome!

>  This isn't as easy as "run on the system OpenGL". 
Yeah I know what you mean, I wrote an HTML canvas API on top of WebGL to support gl-react-native.

On windows any electron app would benefit of this out of the box, right?
With chromium running Angle I have tried magenta style transfer model and seems to work.
In fact setting the --use-gl=desktop option I could see that the chromium engine fallback to Google SwiftShader(CPU) on my machine, looks like my Intel 4400 does not apply for openGL GPU.

After some tests I can see that tfjs cpu performs like 3 seconds worse than Swiftshader and Angle performs three times slower than pure CPU bindings...

3.6secs(Angle) vs 41.6secs(Google SwiftShader) vs 44 secs (web CPU) vs 1.1secs(REAL CPU)

I was expecting Angle to perform at least equal to the real CPU bindings... I'm testing this days on a proper desktop with a medium sized graphic card.
